trytond/doc/topics/configuration.rst
.. _topics-configuration:

==================
Configuration file
==================

The configuration file controls some aspects of the behavior of Tryton.
The file uses a simple ini-file format. It consists of sections, led by a
``[section]`` header and followed by ``name = value`` entries:

.. highlight:: ini

::

    [database]
    uri = postgresql://user:password@localhost/
    path = /var/lib/trytond

For more information see ConfigParser_.

.. _ConfigParser: http://docs.python.org/2/library/configparser.html

The default value of any option can be changed using environment variables
with names using this syntax: ``TRYTOND_<SECTION>__<NAME>``.

Sections
========

This section describes the different main sections that may appear in a Tryton
configuration file, the purpose of each section, its possible keys, and their
possible values.
Some modules could request the usage of other sections for which the guideline
asks them to be named like their module.

.. contents::
   :local:
   :backlinks: entry
   :depth: 2

.. _config-web:

web
---

Defines the behavior of the web interface.

.. _config-web.listen:

listen
~~~~~~

Defines the couple of host (or IP address) and port number separated by a colon
to listen on.

Default ``localhost:8000``

.. note::
   To listen on all IPv4 interfaces use the value ``0.0.0.0:8000`` and for all
   IPv6 interfaces use ``[::]:8000``.

.. _config-web.hostname:

hostname
~~~~~~~~

Defines the hostname to use when generating a URL when there is no request
context available, for example during a cron job.

.. _config-web.root_path:

root_path
~~~~~~~~~

Defines the prefix that the WSGI application is mounted under (with trailing slash).
This must be the same as the ``SCRIPT_NAME`` set by the WSGI server.

The default value is: ``/``.

.. _config-web.root:

root
~~~~

Defines the root path served by ``GET`` requests.

Default: Under the ``www`` directory of user's home running ``trytond``.

.. _config-web.num_proxies:

num_proxies
~~~~~~~~~~~

The number of proxy servers in front of ``trytond``.

Default: 0

.. _config-web.cache_timeout:

cache_timeout
~~~~~~~~~~~~~

The cache timeout in seconds.

Default: 12h

.. _config-web.cors:

cors
~~~~

The list (one per line) of origins allowed for `Cross-Origin Resource sharing
<https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.
For example::

   cors =
      http://example.com
      https://example.com

.. _config-web.avatar_base:

avatar_base
~~~~~~~~~~~

The base URL without a path for avatar URL.

Default: ``''``

.. note:: It can be used to setup a CDN.


.. _config-web.avatar_timeout:

avatar_timeout
~~~~~~~~~~~~~~

The time in seconds that the avatar can be stored in cache.

Default: 7 days

.. _config-database:

database
--------

Defines how the database is managed.

.. _config-database.uri:

uri
~~~

Contains the URI to connect to the SQL database. The URI follows the :rfc:`3986`.
The typical form is:

    database://username:password@host:port/?param1=value1&param2=value2

The parameters are database dependent, check the database documentation for a
list of valid parameters.

Default: The value of the environment variable ``TRYTOND_DATABASE_URI`` or
``sqlite://`` if not set.

The available databases are:

PostgreSQL
**********

``psycopg`` supports two type of connections:

- TCP/IP connection: ``postgresql://user:password@localhost:5432/``
- Unix domain connection:
   - with password authentication: ``postgresql://username:password``
   - with operating system user name: ``postgresql://``

Please refer to `psycopg for the complete specification of the URI
<https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.connect>`_.

A list of parameters supported by PostgreSQL can be found in the
`documentation
<https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS>`__.

.. note::
   ``fallback_application_name`` parameter from aforementioned documentation can
   be set directly thanks to the ``TRYTOND_APPNAME`` environment variable.

SQLite
******

The URI is defined as ``sqlite://``

If the name of the database is ``:memory:``, the parameter ``mode`` will be set
to ``memory`` thus using a pure in-memory database.

The recognized query parameters can be found in SQLite's
`documentation
<https://www.sqlite.org/uri.html#recognized_query_parameters>`__.

.. _config-database.path:

path
~~~~

The directory where Tryton stores files and so the user running
:command:`trytond` must have write access on this directory.

Default: The :file:`db` folder under the user home directory running
:command:`trytond`.

.. _config-database.list:

list
~~~~

A boolean value to list available databases.

Default: ``True``

.. _config-database.retry:

retry
~~~~~

The number of retries when a database operational error occurs during a request.

Default: ``5``

.. _config-database.subquery_threshold:

subquery_threshold
~~~~~~~~~~~~~~~~~~

The number of records in the target relation under which a sub-query is used.

Default: ``1000``

.. _config-database.language:

language
~~~~~~~~

The main language of the database that will be used for storage in the main
table for translations.

Default: ``en``

.. _config-database.avatar_filestore:

avatar_filestore
~~~~~~~~~~~~~~~~

This configuration value indicates whether the avatars should be stored in the
:py:mod:`trytond.filestore` (``True``) or the database (``False``).

Default: ``False``

.. _config-database.avatar_prefix:

avatar_prefix
~~~~~~~~~~~~~

The prefix to use with the :ref:`FileStore <ref-filestore>` to store avatars.

Default: ``None``

.. _config-database.default_name:

default_name
~~~~~~~~~~~~

The name of the database to use for operations without a database name.
Default: ``template1`` for PostgreSQL, ``:memory:`` for SQLite.

.. _config-database.timeout:

timeout
~~~~~~~

The timeout duration in seconds after which the connections to unused databases
are closed.
Default: ``1800`` (30 minutes)

.. _config-database.minconn:

minconn
~~~~~~~

The minimum number of connections to keep in the pool (if the backend supports
pool) per process.
Default: ``1``

.. _config-database.maxconn:

maxconn
~~~~~~~

The maximum number of simultaneous connections to the database per process.
Default: ``64``

.. _config-database.unaccent_function:

unaccent_function
~~~~~~~~~~~~~~~~~

The name of the unaccent function.

Default: ``unaccent``

.. _config-database.similarity_function:

similarity_function
~~~~~~~~~~~~~~~~~~~

The name of the similarity function.

Default: ``similarity``

binary_scanner
~~~~~~~~~~~~~~

The command used to scan the content of Binary fields.
The command receive the quarantine directory as last argument.
If the command returns a non zero status, it is considered that at least one
file in the directory is malicious.

binary_scanner_directory
~~~~~~~~~~~~~~~~~~~~~~~~

The directory where the binary content are stored before being checked
by the scanner.

Default: The temporary directory as determined by Python's tempfile_ module.

.. _tempfile: https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir

.. _config-request:

request
-------

.. _config-request.max_size:

max_size
~~~~~~~~

The maximum size in bytes of unauthenticated request (zero means no limit).

Default: 2MB

.. _config-request.max_size_authenticated:

max_size_authenticated
~~~~~~~~~~~~~~~~~~~~~~

The maximum size in bytes of an authenticated request (zero means no limit).

Default: 2GB

.. _config-request.timeout:

timeout
~~~~~~~

The timeout in seconds before aborting requests that have their execution time
depending on the parameters.

Default: ``60``

.. _config-request.records_limit:

records_limit
~~~~~~~~~~~~~

The maximal number of records processed by requests.

Default: ``None``

.. _config-cache:

cache
-----

Defines size of various cache.

.. _config-cache.transaction:

transaction
~~~~~~~~~~~

The number of contextual caches kept per transaction.

Default: ``10``

.. _config-cache.model:

model
~~~~~

The number of different model kept in the cache per transaction.

Default: ``200``

.. _config-cache.record:

record
~~~~~~

The number of record loaded kept in the cache of the list.
It can be changed locally using the ``_record_cache_size`` key in
:attr:`Transaction.context <trytond.transaction.Transaction.context>`.

Default: ``2000``

.. _config-cache.field:

field
~~~~~

The number of field to load with an ``eager`` :attr:`Field.loading
<trytond.model.fields.Field.loading>`.

Default: ``100``

.. _config-cache.default:

default
~~~~~~~

The default :attr:`~trytond.cache.Cache.size_limit` of :class:`~trytond.cache.Cache`.

Default: ``1024``

.. _config-cache.clean_timeout:

clean_timeout
~~~~~~~~~~~~~

The minimum number of seconds between two cleanings of the cache.
If the value is 0, the notification between processes will be done using
channels if the back-end supports them.

Default: ``300``

.. _config-cache.select_timeout:

select_timeout
~~~~~~~~~~~~~~

The timeout duration of the select call when using channels.

Default: ``60``

.. _config-cache.count_timeout:

count_timeout
~~~~~~~~~~~~~

The cache timeout duration in seconds of the estimation of records.

Default: ``86400`` (1 day)

.. _config-cache.count_clear:

count_clear
~~~~~~~~~~~

The number of operations after which the counting estimation of records is
cleared.

Default: ``1000``

.. _config-cron:

cron
----

.. _config-cron.clean_days:

log_size
~~~~~~~~

The approximate number of log entries to keep for each scheduled task.

Default: ``100``

.. _config-queue:

queue
-----

.. _config-queue.worker:

worker
~~~~~~

Activate asynchronous processing of the tasks. Otherwise they are performed at
the end of the requests.

Default: ``False``

.. _config-queue.clean_days:

clean_days
~~~~~~~~~~

The number of days after which processed tasks are removed.

Default: ``30``

.. _config-queue.batch_size:

batch_size
~~~~~~~~~~

The default number of the instances to process in a batch.

Default: ``20``

.. _config-error:

error
-----

.. _config-error.clean_days:

clean_days
~~~~~~~~~~

The number of days after which reported errors are removed.

Default: ``90``

.. _config-table:

table
-----

This section allows to override the default generated table name for a
:class:`~trytond.model.ModelSQL`.
The main goal is to bypass limitation on the name length of the database
backend.
For example::

    [table]
    account.invoice.line = acc_inv_line
    account.invoice.tax = acc_inv_tax

.. _config-table_query_materialized:

table_query_materialized
------------------------

This section allow to define which :class:`~trytond.model.ModelSQL` with
:meth:`~trytond.model.ModelSQL.table_query` must be materialized and at which
interval they must be refreshed.

The interval value is expressed in seconds.

For example::

    [table_query_materialized]
    account.account.party = 1440

.. note::

   Only the :meth:`~trytond.model.ModelSQL.table_query` that does not depend on
   the :attr:`~trytond.transaction.Transaction.context`.

.. _config-ssl:

ssl
---

Activates SSL_ on the web interface.

.. note:: It is recommended to delegate the SSL support to a proxy.

.. _config-ssl.privatekey:

privatekey
~~~~~~~~~~

The path to the private key.

.. _config-ssl.certificate:

certificate
~~~~~~~~~~~

The path to the certificate.

.. tip::
   Set only one of ``privatekey`` or ``certificate`` to ``true`` if the SSL is
   delegated.

.. _config-email:

email
-----

.. note:: Email settings can be tested with the ``trytond-admin`` command

.. _config-email.uri:

uri
~~~

The SMTP-URL_ to connect to the SMTP server which is extended to support SSL_
and STARTTLS_.
The available protocols are:

    - ``smtp``: simple SMTP
    - ``smtp+tls``: SMTP with STARTTLS
    - ``smtps``: SMTP with SSL

The uri accepts the following additional parameters:

* ``local_hostname``: used as FQDN of the local host in the HELO/EHLO commands,
  if omited it will use the value of ``socket.getfqdn()``.
* ``timeout``: A number of seconds used as timeout for blocking operations. A
  ``socket.timeout`` will be raised when exceeded. If omited the default timeout
  will be used.


Default: ``smtp://localhost:25``

.. _config-email.from:

from
~~~~

Defines the default ``From`` address (using :rfc:`5322`) for emails sent by
Tryton.

For example::

    from: "Company Inc" <info@example.com>

Default: The login name of the :abbr:`OS (Operating System)` user.

.. _config-email.retry:

retry
~~~~~

The number of retries when the SMTP server returns a temporary error.

Default: ``5``

.. _config-session:

session
-------

.. _config-session.authentications:

authentications
~~~~~~~~~~~~~~~

A comma separated list of the authentication methods to try when attempting to
verify a user's identity. Each method is tried in turn, following the order of
the list, until one succeeds. In order to allow `multi-factor authentication`_,
individual methods can be combined together using a plus (``+``) symbol.

Example::

    authentications = password+sms,ldap

Each combined method can have options to skip them if they are met except for
the first method.
They are defined by appending their name to the method name after a question
mark (``?``) and separated by colons (``:``).

Example::

   authentications = password+sms?ip_address:device_cookie


By default, Tryton only supports the ``password`` method.  This method compares
the password entered by the user against a stored hash of the user's password.
By default, Tryton supports the ``ip_address`` and ``device_cookie`` options.
The ``ip_address`` compares the client IP address with the known network list
defined in `authentication_ip_network`_.
The ``device_cookie`` checks the client device is a known device of the user.
Other modules can define additional authentication methods and options, please
refer to their documentation for more information.

Default: ``password``

.. _config-session.authentication_ip_network:

authentication_ip_network
~~~~~~~~~~~~~~~~~~~~~~~~~

A comma separated list of known IP networks used to check for ``ip_address``
authentication method option.

Default: ``''``

.. _config-session.cookie_domain:

cookie_domain
~~~~~~~~~~~~~

The cookie `Domain attribute`_ set when the authentication completes.

Default: ``''``

.. note::
   When accessing the server through ``localhost`` this option should be left
   empty as cookies with the domain ``localhost`` are not strored by browsers.

.. _config-session.max_age:

max_age
~~~~~~~

The time in seconds that a session stay valid.

Default: ``2592000`` (30 days)

.. _config-session.timeout:

timeout
~~~~~~~

The time in seconds without activity before the session is no more fresh.

Default: ``300`` (5 minutes)

.. _config-session.max_attempt:

max_attempt
~~~~~~~~~~~

The maximum authentication attempt before the server answers unconditionally
``Too Many Requests`` for any other attempts. The counting is done on all
attempts over a period of ``timeout``.

Default: ``5``

.. _config-session.max_attempt_ip_network:

max_attempt_ip_network
~~~~~~~~~~~~~~~~~~~~~~

The maximum authentication attempt from the same network before the server
answers unconditionally ``Too Many Requests`` for any other attempts. The
counting is done on all attempts over a period of ``timeout``.

Default: ``300``

.. _config-session.ip_network_4:

ip_network_4
~~~~~~~~~~~~

The network prefix to apply on IPv4 address for counting the authentication
attempts.

Default: ``32``

.. _config-session.ip_network_6:

ip_network_6
~~~~~~~~~~~~

The network prefix to apply on IPv6 address for counting the authentication
attempts.

Default: ``56``

.. _config-password:

password
--------

.. _config-password.length:

length
~~~~~~

The minimal length required for the user password.

Default: ``8``

.. _config-password.forbidden:

forbidden
~~~~~~~~~

The path to a file containing one forbidden password per line.

.. _config-password.reset_timeout:

reset_timeout
~~~~~~~~~~~~~

The time in seconds until the reset password expires.

Default: ``86400`` (24h)

.. _config-attachment:

attachment
----------

Defines how to store the attachments

.. _config-attachment.filestore:

filestore
~~~~~~~~~

A boolean value to store attachment in the :ref:`FileStore <ref-filestore>`.

Default: ``True``

.. _config-attachment.store_prefix:

store_prefix
~~~~~~~~~~~~

The prefix to use with the ``FileStore``.

Default: ``None``

.. _config-attachment.retention_delay:

retention_days
~~~~~~~~~~~~~~

The number of days before unused files are actually removed from the
:ref:`FileStore <ref-filestore>`.

Default: 90

.. _config-bus:

bus
---

.. _config-bus.allow_subscribe:

allow_subscribe
~~~~~~~~~~~~~~~

A boolean value to allow clients to subscribe to bus channels.

Default: ``False``

.. _config-bus.url_host:

url_host
~~~~~~~~

If set redirects bus requests to the host URL.

.. _config-bus.long_polling_timeout:

long_polling_timeout
~~~~~~~~~~~~~~~~~~~~

The time in seconds to keep the connection to the client opened when using long
polling for bus messages

Default: ``300``

.. _config-bus.cache_timeout:

cache_timeout
~~~~~~~~~~~~~

The number of seconds a message should be kept by the queue before being
discarded.

Default: ``300``

.. _config-bus.select_timeout:

select_timeout
~~~~~~~~~~~~~~

The timeout duration of the select call when listening on a channel.

Default: ``5``

.. _config-report:

report
------

.. _config-report.convert_command:

convert_command
---------------

The command to convert document between formats.

The available keywords are:

   - ``%(directory)s``: the temporary working directory
   - ``%(input_format)s``: the format of the file to convert
   - ``%(input_extension)s``: the extension of the file to convert
   - ``%(input_path)s``: the path of the file to convert
   - ``%(output_format)s``: the format to which the file must be converted
   - ``%(output_extension)s``: the extension for the converted file
   - ``%(output_path)s``: the path where the converted file must be written

The command must write the result in ``%(output_path)s``.

.. _config-html:

html
----

.. _config-html.src:

src
~~~

The URL pointing to `TinyMCE <https://www.tiny.cloud/>`_ editor.

Default: ``https://cloud.tinymce.com/stable/tinymce.min.js``

.. _config-html.license_key:

license_key
~~~~~~~~~~~

The license key for TinyMCE.

Default: ``gpl``

.. _config-html.plugins:

plugins
~~~~~~~

The space separated list of TinyMCE plugins to load.
It can be overridden for specific models and fields using the names:
``plugins-<model>-<field>`` or ``plugins-<model>``.

Default: ``''``

.. _config-html.css:

css
~~~

The JSON list of CSS files to load.
It can be overridden for specific models and fields using the names:
``css-<model>-<field>`` or ``css-<model>``.

Default: ``[]``

.. _config-html.class:

class
~~~~~

The class to add on the body.
It can be overridden for specific models and fields using the names:
``class-<model>-<field>`` or ``class-<model>``.

Default: ``''``

.. _config-wsgi_middleware:

wsgi middleware
---------------

The section lists the `WSGI middleware`_ class to load.
Each middleware can be configured with a section named ``wsgi <middleware>``
containing ``args`` and ``kwargs`` options.

Example::

    [wsgi middleware]
    ie = werkzeug.contrib.fixers.InternetExplorerFix

    [wsgi ie]
    kwargs={'fix_attach': False}

.. note::
   The options can be set using environment variables with names like:
   ``TRYTOND_WSGI_<MIDDLEWARE>__<NAME>``.


.. _JSON-RPC: http://en.wikipedia.org/wiki/JSON-RPC
.. _XML-RPC: http://en.wikipedia.org/wiki/XML-RPC
.. _SMTP-URL: https://datatracker.ietf.org/doc/html/draft-earhart-url-smtp-00
.. _SSL: http://en.wikipedia.org/wiki/Secure_Sockets_Layer
.. _STARTTLS: http://en.wikipedia.org/wiki/STARTTLS
.. _WSGI middleware: https://en.wikipedia.org/wiki/Web_Server_Gateway_Interface#Specification_overview
.. _`multi-factor authentication`: https://en.wikipedia.org/wiki/Multi-factor_authentication
.. _`Domain attribute`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Set-Cookie#domaindomain-value


sao/CHANGELOG
* Fix the csv export of dict fields
* Don't reset date/datetime/time to None when it is invalid

Version 7.6.6 - 2025-09-03
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.5 - 2025-08-01
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.4 - 2025-07-15
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.3 - 2025-07-01
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.2 - 2025-06-04
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.1 - 2025-05-15
--------------------------
* Bug fixes (see mercurial logs for details)


Version 7.6.0 - 2025-04-28
--------------------------
* Bug fixes (see mercurial logs for details)
* Allow column widths to be reset
* Allow column resizing
* Add simple chat widget on sidebar
* Support color as type of icon and image
* Support color widget on form view

Version 7.4.0 - 2024-11-04
--------------------------
* Bug fixes (see mercurial logs for details)
* Add restore button on Many2Many
* Keep CSV export window opened
* Remove favorite management
* Retry JSON-RPC requests on Service Unavailable
* Allow resizing the attachment preview panel
* Set focus on search entry after search
* Support language suffix in CSV export
* Close tab with mouse middle click
* Support language suffix in CSV import

Version 7.2.0 - 2024-04-29
--------------------------
* Bug fixes (see mercurial logs for details)
* Read xxx2Many fields using the dotted notation
* Display XML ID on Log window
* Use operators when converting PYSON to string
* Add DnD on binary and image widgets
* Allow resetting forgotten password

Version 7.0.0 - 2023-10-30
--------------------------
* Bug fixes (see mercurial logs for details)
* Support create record from autocomplete
* Use grid layout for form container
* Support scanning code
* Reset search offset when domain has changed
* Add a canonicalize function for domains
* Add menu to copy selected record from tree view to clipboard
* Support PYSON comparison of timedelta
* Support encoding timedelta into PYSON TimeDelta
* Eagerly read string value of multiselection fields
* Use locale timezone for CSV in locale format

Version 6.8.0 - 2023-05-01
--------------------------
* Bug fixes (see mercurial logs for details)
* Include events to logs
* Support scroll time on calendar view
* Save and Restore tree state on List Form
* Do record (pre)validation and saving on selection change for list-form views
* Eagerly read string value of selection fields
* Do not select first record by default
* Use ``&`` and ``|`` as boolean operator in search widget
* Use toggle button and label on translate dialog
* Support border type of images
* Manage domain on id in single value list as unique
* Do not validate domain nor enforce uniqueness for empty fields unless they're required or invisible
* Display the number of selected records on xxx2Many

Version 6.6.0 - 2022-10-31
--------------------------
* Bug fixes (see mercurial logs for details)
* Support authentication services
* Check server version

Version 6.4.0 - 2022-05-02
* Bug fixes (see mercurial logs for details)
* Support notification message
* Allow Reference to be edited from tree view
* Set ECMAScript version to 6
* Manage optional column
* Display the number of selected records
* Humanize the count result
* Add limit to search_count
* Call view_get for board view
* Limit board action domain to active id and ids
* Manage creatable attribute of view

Version 6.2.0 - 2021-11-01
* Bug fixes (see mercurial logs for details)
* Support grouping attribute
* Add expander on tree to fold or unfold selected rows
* Support monetary and symbol field attributes
* Support digits from relation field
* Add accesskey to widgets and buttons
* Support Binary field in CSV import/export

Version 6.0.0 - 2021-05-03
* Bug fixes (see mercurial logs for details)
* Use appropriate types for email and url widgets
* Add completion on xxx2Many and Dict
* Add attachment preview
* Support document widget
* Support icon of type URL
* Allow sending email with existing attachments
* Add paths and model context to action data
* Add indentation in CSV export of tree
* Manage which records to use for actions
* Add drag and drop on CSV export and import
* Remove tempusdominus dependency
* Handle device cookie
* Add breadcrumb as title of window form
* Manage help for each selection
* Display revision on dialog
* Support empty value for timedelta converter
* Unify PYSON string format

Version 5.8.0 - 2020-11-02
* Bug fixes (see mercurial logs for details)
* Format timedelta in CSV export
* Manage symbol widget
* Fallback to model name as title
* Set treeview footer sticky on bottom
* Add get, in_ and contains methods to PYSON
* Manage deletable and writable state
* Support email template
* Always encode CSV export in UTF-8
* Use tempusdominus and Popper for date time picker
* Support PYSON comparison of date and datetime
* Customize bootstrap default style
* Set field's name as input's name attribute
* Position copied records based on order
* Sanitize translated RichtText fields content (issue9453)
* Allow configuration of default colors for graph and calendar
* Sanitize RichtText fields content (issue9405)
* Escape external string (issue9394)
* Keep context in sessionStorage
* Use existing context for get_preferences
* Escape external strings (issue9351)
* Add context to export URL
* Allow changes to editable tree checkboxes without first selecting the line

Version 5.6.0 - 2020-05-04
* Bug fixes (see mercurial logs for details)
* Support link button on form
* Add URL for the current export
* Add option to export listed records
* Set treeview header sticky on top
* Add noreferrer noopener to external URL (issue9089)
* Manage readonly dates in calendar view
* Manage editable on calendar view
* Manage named separator as label
* Add MultiSelection entry to Dict field

trytond/trytond/config.py
# This file is part of Tryton.  The COPYRIGHT file at the top level of
# this repository contains the full copyright notices and license terms.
import configparser
import logging
import os
import sys
import urllib.parse
import warnings
from functools import cache
from getpass import getuser

import __main__ as main

__all__ = [
    'get_hostname', 'get_port', 'split_netloc',
    'parse_listen', 'parse_uri',
    'update_etc', 'has_section', 'add_section', 'remove_section',
    'set', 'get', 'getint', 'getfloat', 'getboolean']
logger = logging.getLogger(__name__)

# Needed so urlunsplit to always set netloc
for backend_name in ['postgresql', 'sqlite']:
    if backend_name not in urllib.parse.uses_netloc:
        urllib.parse.uses_netloc.append(backend_name)


def get_hostname(netloc):
    if '[' in netloc and ']' in netloc:
        return netloc.split(']')[0][1:]
    elif ':' in netloc:
        return netloc.split(':')[0]
    else:
        return netloc


def get_port(netloc):
    netloc = netloc.split(']')[-1]
    return int(netloc.split(':')[1])


def split_netloc(netloc):
    return get_hostname(netloc).replace('*', ''), get_port(netloc)


def parse_listen(value):
    for netloc in value.split(','):
        yield split_netloc(netloc)


def parse_uri(uri):
    return urllib.parse.urlparse(uri)


class TrytonConfigParser(configparser.ConfigParser):

    def __init__(self, overrides=None):
        super().__init__(interpolation=None)
        self.add_section('web')
        self.set('web', 'listen', 'localhost:8000')
        self.set('web', 'root_path', '/')
        self.set('web', 'root', os.path.join(os.path.expanduser('~'), 'www'))
        self.set('web', 'num_proxies', '0')
        self.set('web', 'cache_timeout', str(60 * 60 * 12))
        self.add_section('database')
        self.set('database', 'uri',
            os.environ.get('TRYTOND_DATABASE_URI') or 'sqlite://')
        self.set('database', 'path', os.path.join(
                os.path.expanduser('~'), 'db'))
        self.set('database', 'list', 'True')
        self.set('database', 'retry', '5')
        self.set('database', 'language', 'en')
        self.set('database', 'timeout', str(30 * 60))
        self.set('database', 'subquery_threshold', str(1_000))
        self.add_section('request')
        self.set('request', 'max_size', str(2 * 1024 * 1024))
        self.set('request', 'max_size_authenticated',
            str(2 * 1024 * 1024 * 1024))
        self.set('request', 'timeout', str(60))
        self.add_section('cache')
        self.set('cache', 'transaction', '10')
        self.set('cache', 'model', '200')
        self.set('cache', 'record', '2000')
        self.set('cache', 'field', '100')
        self.set('cache', 'default', '1024')
        self.set('cache', 'ir.message', '10240')
        self.set('cache', 'ir.translation', '10240')
        self.set('cache', 'clean_timeout', '300')
        self.set('cache', 'select_timeout', '60')
        # AKE: cache config from env vars
        self.set('cache', 'class', os.environ.get('TRYTOND_CACHE_CLASS', ''))
        self.set('cache', 'uri', os.environ.get('TRYTOND_CACHE_URI', ''))
        self.set('cache', 'coog_cache_size', '1024')

        self.add_section('report')
        self.set('report', 'unoconv_retry', '2')

        self.add_section('queue')
        self.set('queue', 'worker', 'False')
        self.add_section('ssl')
        self.add_section('email')
        self.set('email', 'uri', 'smtp://localhost:25')
        self.set('email', 'from', getuser())
        self.add_section('session')
        self.set('session', 'authentications', 'password')
        self.set('session', 'max_age', str(60 * 60 * 24 * 30))
        self.set('session', 'timeout', str(60 * 5))
        self.set('session', 'max_attempt', '5')
        self.set('session', 'max_attempt_ip_network', '300')
        self.set('session', 'ip_network_4', '32')
        self.set('session', 'ip_network_6', '56')
        self.add_section('password')
        self.set('password', 'length', '8')
        self.set('password', 'reset_timeout', str(24 * 60 * 60))
        self.add_section('bus')
        self.set('bus', 'allow_subscribe', 'False')
        self.set('bus', 'long_polling_timeout', str(5 * 60))
        self.set('bus', 'cache_timeout', '5')
        self.set('bus', 'select_timeout', '5')
        self.add_section('html')
        self.add_section('attachment')
        self.add_section('custom')
        self.set('custom', 'enable_stat_thread', 'False')
        if overrides:
            self.update_etc(configfile=overrides)
        self.update_environ()
        self.update_etc()

    def update_environ(self):
        for key, value in os.environ.items():
            if not key.startswith('TRYTOND_'):
                continue
            try:
                section, option = key[len('TRYTOND_'):].lower().split('__', 1)
            except ValueError:
                continue
            if section.startswith('wsgi_'):
                section = section.replace('wsgi_', 'wsgi ')
            if section.startswith('authentication_saml'):
                section = section.replace(
                    'authentication_saml_', 'authentication_saml ')
            if not self.has_section(section):
                self.add_section(section)
            self.set(section, option, value)

    def update_etc(self, configfile=os.environ.get('TRYTOND_CONFIG')):
        if not configfile:
            return []
        if isinstance(configfile, str):
            configfile = [configfile]
        configfile = [
            os.path.expanduser(filename) for filename in configfile
            if filename]
        if not configfile:
            return []
        read_files = self.read(configfile)
        logger.info('using %s as configuration files', ', '.join(read_files))
        if configfile != read_files:
            logger.error('could not load %s',
                ','.join(set(configfile) - set(read_files)))
        return configfile

    def apply_overriden_defaults(self):
        overrides = self.get(
            'admin', 'config_default_overrides', fallback=False)
        if not overrides:
            return self
        return TrytonConfigParser(overrides=overrides)


_config = TrytonConfigParser()
_config.apply_overriden_defaults()


def update_etc(configfile=os.environ.get('TRYTOND_CONFIG')):
    configfile = _config.update_etc(configfile=configfile)
    _cache_clear()
    return configfile


has_section = _config.has_section
add_section = _config.add_section
remove_section = _config.remove_section
options = _config.options


def set(section, option, value=None):
    _config.set(section, option, value=value)
    _cache_clear()


@cache
def get(section, option, default=None):
    return configparser.RawConfigParser.get(
        _config, section, option, fallback=default)


def _cache_clear():
    get.cache_clear()
    getint.cache_clear()
    getfloat.cache_clear()
    getboolean.cache_clear()


@cache
def getint(section, option, default=None):
    return configparser.RawConfigParser.getint(
        _config, section, option, fallback=default)


@cache
def getfloat(section, option, default=None):
    return configparser.RawConfigParser.getfloat(
        _config, section, option, fallback=default)


@cache
def getboolean(section, option, default=None):
    return configparser.RawConfigParser.getboolean(
        _config, section, option, fallback=default)


def __getattr__(name):
    if name == 'config':
        warnings.warn(
            "trytond.config.config is deprecated, "
            "use trytond.config", DeprecationWarning, stacklevel=2)
        return sys.modules[__name__]
    raise AttributeError(f"module {__name__} has no attribute {name}")
